src/util/sample.cpp

```diff
@@ -527,33 +568,202 @@ void SampleUtil::linearCrossfadeBuffersOut(
+        pDestSrcFadeIn[i * 8 + 1] += pSrcFadeOut[i * 8 + 1] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+    for (int i = 0; i < numSamples / 8; ++i) {
+        const CSAMPLE_GAIN cross_mix = cross_inc * i;
+        pDestSrcFadeIn[i * 8 + 2] *= cross_mix;
+        pDestSrcFadeIn[i * 8 + 2] += pSrcFadeOut[i * 8 + 2] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+    for (int i = 0; i < numSamples / 8; ++i) {
+        const CSAMPLE_GAIN cross_mix = cross_inc * i;
+        pDestSrcFadeIn[i * 8 + 3] *= cross_mix;
+        pDestSrcFadeIn[i * 8 + 3] += pSrcFadeOut[i * 8 + 3] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+    for (int i = 0; i < numSamples / 8; ++i) {
+        const CSAMPLE_GAIN cross_mix = cross_inc * i;
+        pDestSrcFadeIn[i * 8 + 4] *= cross_mix;
+        pDestSrcFadeIn[i * 8 + 4] += pSrcFadeOut[i * 8 + 4] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+    for (int i = 0; i < numSamples / 8; ++i) {
+        const CSAMPLE_GAIN cross_mix = cross_inc * i;
+        pDestSrcFadeIn[i * 8 + 5] *= cross_mix;
+        pDestSrcFadeIn[i * 8 + 5] += pSrcFadeOut[i * 8 + 5] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+    for (int i = 0; i < numSamples / 8; ++i) {
+        const CSAMPLE_GAIN cross_mix = cross_inc * i;
+        pDestSrcFadeIn[i * 8 + 6] *= cross_mix;
+        pDestSrcFadeIn[i * 8 + 6] += pSrcFadeOut[i * 8 + 6] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+    for (int i = 0; i < numSamples / 8; ++i) {
+        const CSAMPLE_GAIN cross_mix = cross_inc * i;
+        pDestSrcFadeIn[i * 8 + 7] *= cross_mix;
+        pDestSrcFadeIn[i * 8 + 7] += pSrcFadeOut[i * 8 + 7] * (CSAMPLE_GAIN_ONE - cross_mix);
+    }
+}
+
+// static
+void SampleUtil::linearCrossfadeBuffersIn(
+        CSAMPLE* pDestSrcFadeIn,
+        const CSAMPLE* pSrcFadeOut,
+        SINT numSamples,
+        int channelCount) {
+    switch (channelCount) {
+    case mixxx::audio::ChannelCount::stereo():
+        return SampleUtil::linearCrossfadeStereoBuffersIn(
+                pDestSrcFadeIn,
+                pSrcFadeOut,
+                numSamples);
+    case mixxx::audio::ChannelCount::stem():
+        return SampleUtil::linearCrossfadeStemBuffersIn(
+                pDestSrcFadeIn,
+                pSrcFadeOut,
+                numSamples);
+    default:
+        // Fallback to unoptimised function
+        {
+            int numFrame = numSamples / channelCount;
+            const CSAMPLE_GAIN cross_inc =
+                    CSAMPLE_GAIN_ONE / CSAMPLE_GAIN(numSamples / channelCount);
+            for (int chIdx = 0; chIdx < channelCount; chIdx++) {
+                for (int i = 0; i < numFrame; ++i) {
+                    const CSAMPLE_GAIN cross_mix = cross_inc * i;
+                    pDestSrcFadeIn[i * channelCount + chIdx] *= cross_mix;
+                    pDestSrcFadeIn[i * channelCount + chIdx] +=
+                            pSrcFadeOut[i * channelCount + chIdx] *
+                            (CSAMPLE_GAIN_ONE - cross_mix);
+                }
+            }
```

Comment:
Looking at `SampleUtil::linearCrossfadeStemBuffersIn`, it was containing 8 loops instead of a single one. I believe the specialisation should now be feaster due to vectorization. Could you test it again, so we get consistent hardware to benchmark on?

Reply:
Wow, It is blazing fats now: 
BM_LinearCrossfadeBuffersOut/64           16,6 ns         16,6 ns     41604160
BM_LinearCrossfadeBuffersOut/512           121 ns          121 ns      5735540
BM_LinearCrossfadeBuffersOut/4096          978 ns          978 ns       747782

Reply:
The same trick fails totally for stereo. Can you apply the same trick for 'SampleUtil::linearCrossfadeStemBuffersOut' ?

Reply:
Done!